node/src/exchange/server.rs

```diff
@@ -1,52 +1,382 @@
+use crate::exchange::{ReqRespBehaviour, ResponseType};
+use crate::executor::spawn;
 use crate::store::Store;
 
-pub(super) struct ExchangeServerHandler<S>
+pub(super) struct ExchangeServerHandler<S, R = ReqRespBehaviour>
 where
-    S: Store + 'static,
+    S: Store,
+    R: ResponseSender,
 {
-    _store: Arc<S>,
+    store: Arc<S>,
+
+    rx: mpsc::Receiver<(R::Channel, ResponseType)>,
+    tx: mpsc::Sender<(R::Channel, ResponseType)>,
+}
+
+pub(super) trait ResponseSender {
+    type Channel: Send + 'static;
+
+    fn send_response(&mut self, channel: Self::Channel, response: ResponseType);
 }
 
-impl<S> ExchangeServerHandler<S>
+impl ResponseSender for ReqRespBehaviour {
+    type Channel = ResponseChannel<ResponseType>;
+
+    fn send_response(&mut self, channel: Self::Channel, response: ResponseType) {
+        // response was prepared specifically for the request, we can drop it
+        // in case of error we'll get Event::InboundFailure
+        let _ = self.send_response(channel, response);
+    }
+}
+
+impl<S, R> ExchangeServerHandler<S, R>
 where
     S: Store + 'static,
+    R: ResponseSender,
 {
     pub(super) fn new(store: Arc<S>) -> Self {
-        ExchangeServerHandler { _store: store }
+        let (tx, rx) = mpsc::channel(32);
+        ExchangeServerHandler { store, rx, tx }
     }
 
-    #[instrument(level = "trace", skip(self, _respond_to))]
-    pub(super) fn on_request_received(
+    #[instrument(level = "trace", skip(self, response_channel))]
+    pub(super) fn on_request_received<Id>(
         &mut self,
         peer: PeerId,
-        request_id: RequestId,
+        request_id: Id,
         request: HeaderRequest,
-        _respond_to: ResponseChannel<Vec<HeaderResponse>>,
-    ) {
-        // TODO
+        response_channel: R::Channel,
+    ) where
+        Id: Display + Debug,
+    {
+        let Some((amount, data)) = parse_request(request) else {
+            self.handle_invalid_request(response_channel);
+            return;
+        };
+
+        match data {
+            header_request::Data::Origin(0) => {
+                self.handle_request_current_head(response_channel);
+            }
+            header_request::Data::Origin(height) => {
+                self.handle_request_by_height(response_channel, height, amount);
+            }
+            header_request::Data::Hash(hash) => {
+                self.handle_request_by_hash(response_channel, hash);
+            }
+        };
     }
 
-    #[instrument(level = "trace", skip(self))]
     pub(super) fn on_response_sent(&mut self, peer: PeerId, request_id: RequestId) {
-        // TODO
+        trace!("response_sent; request_id: {request_id}, peer: {peer}");
     }
 
-    #[instrument(level = "trace", skip(self))]
     pub(super) fn on_failure(
         &mut self,
         peer: PeerId,
         request_id: RequestId,
         error: InboundFailure,
     ) {
-        // TODO
+        // TODO: cancel job if libp2p already failed it?
```

Comment:
I think the failure here only happens if `send_response` fails. If that is the case, then job is already done.

Reply:
[`InboundFailure`](https://docs.rs/libp2p/latest/libp2p/request_response/enum.InboundFailure.html) can be sent even if we never call `send_response`. I've added artificial delay to spawned task, so that libp2p times out and I see we get 
```
...  WARN celestia_node::exchange::server: on_failure; request_id: 264, peer: 12D3KooWNJqKMTTH6Tmj5CNNksS1dkcMQzkqfAuoz98gdNQb895U, error: ConnectionClosed
```
But I'm still not sure if tracking necessary to manage jobs here is worth the potential resource save.

Reply:
I think it doesn't worth it in our scenario because we don't have any heavy computation. 

Resources will became an issue when a malicious client is spamming us with requests and the channel gets full, in that case resources will be eaten by huge amount of spawned tasks. We need to somehow restrict on how many requests we can handle per time on a higher level.

Reply:
Maybe we can use this: https://docs.rs/libp2p/latest/libp2p/swarm/struct.SwarmBuilder.html#method.max_negotiating_inbound_streams

Also by dropping the `ResponseChannel` the stream will be closed. So, we may track the number of active requests and drop the rest.

Reply:
created https://github.com/eigerco/celestia-node-rs/issues/90 for future tracking